HTTP basic auth support

svvac opened the pull request:
### Proposed changes
Add support for HTTP Basic authentication to Policy objects and to Ingresses via annotations.

#200 #1872

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto master
- [x] I will ensure my PR is targeting the master branch and pulling from my branch from my own fork

svvac commented:
First draft to get the discussion going regarding naming, etc.

Will add tests and docs.

pleshakov commented on internal/configs/annotations.go:
```diff
@@ -295,6 +298,13 @@ func parseAnnotations(ingEx *IngressEx, baseCfgParams *ConfigParams, isPlus bool
 		}
 	}
 
+	if basicSecret, exists := ingEx.Ingress.Annotations[BasicAuthSecretAnnotation]; exists {
```
the annotations names look good.

Could you add validation of their values here internal/k8s/validation.go ?

nginx.org/basic-auth-realm, of type string, required. validation: (1) must have all '"' (double quotes) escaped and must not end with an unescaped '\' (backslash) (2) must not have any variables.  also requires nginx.org/basic-auth-secret to be set.

nginx.org/basic-auth-secret, of type string, required. reference to the secret within the same namespace. validation: valid secret name. also requires nginx.org/basic-auth-realm to be set.

pleshakov replied:
an update based on the conversation about the realm field of the policy:

nginx.org/basic-auth-realm can be optional and can have value "".  However, it is still cannot be used without nginx.org/basic-auth-secret.

pleshakov reviewed (commented):
Hi @svvac 

Thanks for the pull request! This is a great addition to the Ingress Controller.

I left a few comments, questions and suggestions.  As you mentioned, let's finalize the requirements first, this review doesn't include comments about missing tests, etc.

svvac commented on pkg/apis/configuration/validation/policy.go:
```diff
@@ -177,6 +187,18 @@ func validateJWT(jwt *v1.JWTAuth, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if basic.Realm == "" {
+		return append(allErrs, field.Required(fieldPath.Child("realm"), ""))
```
Correct me if I'm wrong, but `validateJWTRealm` looks like it's applying strict validation rules in order to prevent quoting errors in the NginX configuration.

These should be dealt with in the templates (`{{ sprintf "%q" .Realm }}` instead of `"{{ .Realm }}"`), which has the added benefit of allowing setting arbitrary realm strings with quotes.

Also, I like it being optional.

svvac replied:
Silly, I realize the commented line actually requires a value to be set for the realm ><
Would you mind it being made optional? NginX should be happy with having `""` according to the docs.

pleshakov replied:
> Correct me if I'm wrong, but validateJWTRealm looks like it's applying strict validation rules in order to prevent quoting errors in the NginX configuration.

The validation is `a valid realm must have all '"' escaped and must not contain any '$' or end with an unescaped '\'`  We don't allow $, because $ can indicate a variable. and if the variable doesn't exist, NGINX will fail to reload. For example:
```yaml
realm: "$test"
```
will lead to `nginx: [emerg] unknown \"test\" variable\n`

> These should be dealt with in the templates ({{ sprintf "%q" .Realm }} instead of "{{ .Realm }}"), which has the added benefit of allowing setting arbitrary realm strings with quotes.

thanks for the clarification, I think I get it now. However, in other similar fields, like the realm of the JWT policy, we don't allow unescaped symbols. So I think it would be more consistent if we reuse the existing validation.  So if somebody would want to put `"`, they will need to escape it in the YAML.

> Would you mind it being made optional? NginX should be happy with having "" according to the docs.

Yep. This makes sense to me. If it is empty, NGINX will use ""

svvac commented on pkg/apis/configuration/v1/types.go:
```diff
@@ -352,6 +352,7 @@ type PolicySpec struct {
 	AccessControl *AccessControl `json:"accessControl"`
 	RateLimit     *RateLimit     `json:"rateLimit"`
 	JWTAuth       *JWTAuth       `json:"jwt"`
+	BasicAuth     *BasicAuth     `json:"basic"`
```
I considered it, but went with `basic` to align with JWT not being `jwtAuth`. Tell me either way.

pleshakov replied:
I see your logic - basic is indeed consistent with jwt. However, basic doesn't reflect that it is an auth policy in contrast with jwt, so basicAuth makes more sense to me

svvac commented on internal/k8s/secrets/validation.go:
```diff
@@ -28,6 +31,9 @@ const SecretTypeJWK api_v1.SecretType = "nginx.org/jwk"
 const SecretTypeOIDC api_v1.SecretType = "nginx.org/oidc"
 
+// SecretTypeHtpasswd contains an htpasswd file for use in HTTP Basic authorization.. #nosec G101
+const SecretTypeHtpasswd api_v1.SecretType = "nginx.org/htpasswd"
```
Yeah that's the idea. I considered using keys of the secret as usernames and their value as auth string, but this would add processing and the user wouldn't be able to use existing passwd databases with e.g. comments.

brianehlert replied:
My thought here is the experience of creating the htpasswd file and how that feels bespoke to a Kubernetes (yaml focused) operator.
I understand the choice though.  The file is both native to NGINX configuration and not unusual in itself in the full context.
Documentation might help me here - walking through defining, generating, and using the htpasswd as part of the configuration.
We have tutorials or examples that could be used for that.

ciarams87 replied:
There's a gosec linting error, but it's a false positive. We need to skip the error similar to other places. 
```suggestion
const SecretTypeHtpasswd api_v1.SecretType = "nginx.org/htpasswd" // #nosec G101
```

svvac commented on pkg/apis/configuration/validation/policy.go:
```diff
@@ -177,6 +187,18 @@ func validateJWT(jwt *v1.JWTAuth, fieldPath *field.Path) field.ErrorList {
+		return append(allErrs, field.Required(fieldPath.Child("realm"), ""))
+	}
+
+	allErrs = append(allErrs, validateSecretName(basic.Secret, fieldPath.Child("secret"))...)
```
Yeah, good catch

svvac commented:
I've made the requested changes and I'll start working on the tests.

Ankitchandre commented:
Do we have any timeline , when will the basic auth support available ? we want to use annotations for basic auth in our ingress .

ciarams87 commented:
Hi @svvac - I see the update codegen check is failing. Could you please run `make update-codegen` and commit the changes. Please note that the update codegen script expects to be ran in GOPATH - if it isn't, the generated files will be in the wrong location (see https://github.com/knative/pkg/issues/1287 for a similar issue). You can run `cp -r ../../../github.com/nginxinc/kubernetes-ingress/pkg/ pkg/` after running update codegen to copy the required files if necessary.

svvac commented:
@ciarams87 Wow that stuff is wiered, it leaves files at unexpected places. Should be fixed in latest revision.

Regarding progress: I'm drowning in work these days, so if anyone wants to move this forward by adding tests their help is welcome. The rest of the work should be mostly done.

ciarams87 commented:
> @ciarams87 Wow that stuff is weird, it leaves files at unexpected places. Should be fixed in latest revision.
> 
> Regarding progress: I'm drowning in work these days, so if anyone wants to move this forward by adding tests their help is welcome. The rest of the work should be mostly done.

@svvac Totally understand, thanks so much for all your work on this. We will prioritise this as part of our backlog and move it forward.

svvac commented:
Hi @pleshakov @brianehlert @ciarams87 ! I finally got around to move this forward

- Fixed a template bug (basic auth was not configured for mergeable ingresses)
- Added unit tests
- Added integration tests, thanks to @comrem 
- Added docs
- Rebased against main
  - I _think_ I updated codegen properly during rebase
  - The PR has separate commits for each of Policy support and Ingress annotation support, and then again for each of feature, integration tests, and documentation. Just tell me if you want it arranged differently.

Please review and advise for picking this back up!

svvac commented:
Apologies, I missed the removal of `enablePreviewPolicies` during rebase. Latest should correct.

<details>
<summary>FYI, there seems to be a problem in the <code>lint</code> make rule, I fixed it with the following (not in patch set)</summary>

```diff
diff --git i/Makefile w/Makefile
index d6c15872..584be8b4 100644
--- i/Makefile
+++ w/Makefile
@@ -29,8 +29,8 @@ all: test lint verify-codegen update-crds debian-image

 .PHONY: lint
 lint: ## Run linter
-       @git fetch
-       docker run --pull always --rm -v $(shell pwd):/kubernetes-ingress -w /kubernetes-ingress -v $(shell go env GOCACHE):/cache/go -e GOCACHE=/cache/go -e GOLANGCI_LINT_CACHE=/cache/go -v $(shell go env GOPATH)/pkg:/go/pkg golangci/golangci-lint:latest git diff -p origin/main > /tmp/diff.patch && golangci-lint --color always run -v --new-from-patch=/tmp/diff.patch
+       @git fetch origin
+       docker run --pull always --rm -v $(shell pwd):/kubernetes-ingress -w /kubernetes-ingress -v $(shell go env GOCACHE):/cache/go -e GOCACHE=/cache/go -e GOLANGCI_LINT_CACHE=/cache/go -v $(shell go env GOPATH)/pkg:/go/pkg golangci/golangci-lint:latest sh -c 'git diff -p origin/main > /tmp/diff.patch && golangci-lint --color always run -v --new-from-patch=/tmp/diff.patch'

 .PHONY: test
 test: ## Run tests
```

</details>

lucacome commented:
@svvac thanks for the new changes! Just wanted to let you know that we're taking a look at them 🙂

svvac commented:
Hi, is there anything I can do to help move this forward?

ciarams87 commented on internal/k8s/validation.go:
```diff
@@ -572,6 +583,22 @@ func validateSnippetsAnnotation(context *annotationValidationContext) field.Erro
 
+func validateSecretNameAnnotation(context *annotationValidationContext) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if err := validateIsValidSecretName(context.value); err != nil {
```
Suggestion: There is a kubernetes apimachinery validation function that does the same thing as this new function: https://pkg.go.dev/k8s.io/apimachinery@v0.23.6/pkg/util/validation#IsDNS1123Subdomain

ciarams87 commented on tests/suite/test_auth_basic_auth_mergeable.py:
```diff
@@ -0,0 +1,171 @@
+                            },
+                            allow_redirects=False)
+
+        # FIXME: delete this
```
Gentle reminder on this FIXME 😄

ciarams87 commented on tests/suite/wip_test_auth_basic_auth_mergeable.py:
```diff
@@ -0,0 +1,165 @@
+import requests
```
Is this file meant to still be here? It's marked as WIP and looks at a glance to be covering the functionality of the non-WIP named file above.

ciarams87 commented on tests/suite/wip_test_auth_basic_secrets.py:
```diff
@@ -0,0 +1,117 @@
+import requests
```
Same feedback as above

ciarams87 reviewed (commented):
Hi @svvac - apologies again for the delay. I'm pulling down your code here to run through the example myself, but this overall looks great! I've a couple of minor comments, and there are some merge conflicts that need to be addressed, but otherwise this is definitely nearly there!

Thank you so much for all the work so far, and for your patience.

ciarams87 commented on examples/basic-auth/README.md:
```diff
@@ -0,0 +1,112 @@
+
+NGINX supports authenticating requests with [ngx_http_auth_basic_module](https://nginx.org/en/docs/http/ngx_http_auth_basic_module.html).
+
+The Ingress controller provides the following 4 annotations for configuring JWT validation:
```
```suggestion
The Ingress controller provides the following 2 annotations for configuring Basic Auth validation:
```

ciarams87 commented on examples/basic-auth/README.md:
```diff
@@ -0,0 +1,112 @@
+  rules:
+  - host: cafe.example.com
+    http:
+      paths:
+      - path: /tea
+        backend:
+          serviceName: tea-svc
+          servicePort: 80
+      - path: /coffee
+        backend:
+          serviceName: coffee-svc
+          servicePort: 80
```
```suggestion
       paths:
       - path: /tea
          pathType: Prefix
          backend:
            service:
              name: tea-svc
              port:
                number: 80
        - path: /coffee
          pathType: Prefix
          backend:
            service:
              name: coffee-svc
              port:
                number: 80
```

ciarams87 commented on examples/basic-auth/README.md:
```diff
@@ -0,0 +1,112 @@
+
+## Example 2: a Separate Htpasswd Per Path
+
+In the following example we enable JWT validation for the [mergeable Ingresses](../mergeable-ingress-types) with a separate JWT key per path:
```
```suggestion
In the following example we enable Basic Auth validation for the [mergeable Ingresses](../mergeable-ingress-types) with a separate Basic Auth user:password list per path:
```

svvac commented:
@ciarams87 Good news then! I pushed a rebased version against `main`, including fixes regarding the issues you raised

ciarams87 reviewed (approved):
🚀 LGTM!! Thanks again!!

hostalp commented:
Hello, could you consider changing the secet resource requirement so that it would be possible to use the same secret for both Kubernetes ingress controller as well as for this one?

Their secret looks like the following example: https://kubernetes.github.io/ingress-nginx/examples/auth/basic/#examine-secret
With the 2 most important differences being:

- `type: Opaque` (this ingress controller requires `type: nginx.org\htpasswd`)
- auth data being located in `data.auth` (this ingress controller requires them in `data.htpasswd`)

It would make things easier.

brianehlert commented:
> Hello, could you consider changing the secet resource requirement so that it would be possible to use the same secret for both Kubernetes ingress controller as well as for this one?
> 
> Their secret looks like the following example: https://kubernetes.github.io/ingress-nginx/examples/auth/basic/#examine-secret With the 2 most important differences being:
> 
> * `type: Opaque` (this ingress controller requires `type: nginx.org\htpasswd`)
> * auth data being located in `data.auth` (this ingress controller requires them in `data.htpasswd`)
> 
> It would make things easier.

Since this PR was merged approximately 9 months ago.  Would you consider adding this request to Issues as a proposal?

hostalp commented:
Sure, I just didn't want to open a new issue not knowing if this proposal would be declined anyway.
